frontend/include/chpl/resolution/resolution-types.h

```diff
@@ -84,25 +84,38 @@ enum struct DefaultsPolicy {
  */
 class UntypedFnSignature {
  public:
+  enum DefaultKind {
+    /** Formals that have default values, like `in x = 10` */
+    DK_DEFAULT,
+    /** Formals that do not have default values, like `ref x` */
+    DK_NO_DEFAULT,
+    /** Formals that might have a default value. This comes up when working
+        with generic initializers; whether an initializer's formal has
+        a default depends on if its type has a default value. But if
+        the type is unknown -- as in a generic initializer's type signature --
+        then we don't know if the formal has a default. */
+    DK_MAYBE_DEFAULT,
+  };
+
   struct FormalDetail {
     UniqueString name;
-    bool hasDefaultValue = false;
+    DefaultKind defaultKind = DK_NO_DEFAULT;
     const uast::Decl* decl = nullptr;
     bool isVarArgs = false;
 
     FormalDetail(UniqueString name,
-                 bool hasDefaultValue,
+                 DefaultKind defaultKind,
                  const uast::Decl* decl,
                  bool isVarArgs = false)
       : name(name),
-        hasDefaultValue(hasDefaultValue),
+        defaultKind(defaultKind),
         decl(decl),
         isVarArgs(isVarArgs)
     { }
```

Comment:
I will settle for an assertion which is a more lightweigh approach, but did you mean `DK_MAYBE_DEFAULT`?

i think the idea here is to say "DK_MAYBE_DEFAULT implies `init`", because only in initializers could we be unsure about the default value of an actual.

Reply:
Actually, `name` here is the formal name, so comparing against `init` is bad. I guess I'll make a factory.

Reply:
Oh, I meant to say `"this"`, not `"init"`. I mean to assert `this` implies `DK_NO_DEFAULT` since that's the `defaultKind` always used for the `this` formal.

Reply:
Oh, that's much easier :)